cohort-evaluator/src/main/java/com/ibm/cohort/cql/evaluation/CqlContextFactory.java

```diff
@@ -26,70 +32,206 @@
 public class CqlContextFactory {
 
     /**
-     * Stores the timezone relative datetime that represents "now" during CQL engine
-     * execution. Defaults to ZonedDateTime.now() when not provided.
+     * This class captures all of the data that is included in a CQL context 
+     * object that either must remain the same between use of the context 
+     * because it cannot be changed once set or generally does remain the 
+     * same between evaluations (libraryProvider, terminologyProvider).
      */
-    private ZonedDateTime evaluationDateTime = null;
-
-    /**
-     * Controls whether CQL engine trace logging is enabled. Defaults to false.
-     */
-    private CqlDebug debug = CqlDebug.NONE;
+    protected static class ContextCacheKey {
+        public CqlLibraryProvider libraryProvider;
+        public CqlTerminologyProvider terminologyProvider;
+        public CqlLibraryDescriptor topLevelLibrary;
+        public ZonedDateTime evaluationDateTime;
+        public Map<String,Object> parameters;
```

Comment:
Pedantic: Normally, I'd shy away from throwing `final` suggestions your way, but I think in this case it's extra important that these fields do not change.
Since you already have a full constructor setup, it should be easy to set these as `public final` without issue.